session/session.go

```diff
@@ -11,447 +11,243 @@
-			s.healthcheckTimeout = time.Now().Add(healthcheckInterval + healthcheckTimeout)
-		case <-time.After(time.Until(s.healthcheckTimeout)):
-			if s.healthcheckTimeout.Before(time.Now()) {
-				log.Errorf("session %s, health check failed, connection with the client lost", s.id)
-				s.expiresAt = time.Now()
+		case <-timerPing.C:
+			if sessIdle {
+				sess.Error(&ws.ProtoMsg{}, true, "session timeout")
 				return
+			} else {
+				err := sess.Ping()
+				if err != nil {
+					log.Errorf("failed to ping client: %s", err.Error())
+					return
+				}
+				sessIdle = true
+				timerPing.Reset(pongWait)
 			}
-		case <-time.After(time.Until(nextHealthcheckPing)):
-			s.healthcheckPing()
-			nextHealthcheckPing = time.Now().Add(healthcheckInterval)
-		}
-	}
-}
-
-func (s *MenderShellSession) healthcheckPing() {
-	msg := &ws.ProtoMsg{
-		Header: ws.ProtoHdr{
-			Proto:     ws.ProtoTypeShell,
-			MsgType:   wsshell.MessageTypePingShell,
-			SessionID: s.id,
-			Properties: map[string]interface{}{
-				"timeout": int(healthcheckInterval.Seconds() + healthcheckTimeout.Seconds()),
-				"status":  wsshell.ControlMessage,
-			},
-		},
-		Body: nil,
-	}
-	log.Debugf("session %s healthcheck ping", s.id)
-	err := connectionmanager.Write(ws.ProtoTypeShell, msg)
-	if err != nil {
-		log.Debugf("error on write: %s", err.Error())
-	}
-}
-
-func (s *MenderShellSession) HealthcheckPong() {
-	s.pong <- struct{}{}
-}
-
-func (s *MenderShellSession) ShellCommand(m *ws.ProtoMsg) error {
-	s.activeAt = timeNow()
-	data := m.Body
-	commandLine := string(data)
-	n, err := s.writer.Write(data)
-	if err != nil && n != len(data) {
-		err = shell.ErrExecWriteBytesShort
-	}
-	if err != nil {
-		log.Debugf("error: '%s' while running '%s'.", err.Error(), commandLine)
-	} else {
-		log.Debugf("executed: '%s'", commandLine)
-	}
-	return err
-}
-
-func (s *MenderShellSession) ResizeShell(height, width uint16) {
-	shell.ResizeShell(s.pseudoTTY, height, width)
-}
-
-func (s *MenderShellSession) StopShell() (err error) {
-	log.Infof("session %s status:%d stopping shell", s.id, s.status)
-	if s.status != ActiveSession && s.status != HangedSession {
-		return ErrSessionShellNotRunning
-	}
+			continue
 
-	close(s.stop)
-	s.shell.Stop()
-	s.terminal = MenderShellTerminalSettings{}
-	s.status = EmptySession
+		case msg, open = <-sess.msgChan:
+			if !open {
+				return
+			}
+			timerPing.Reset(sess.Config.IdleTimeout)
+			sessIdle = false
+		}
 
-	p, err := os.FindProcess(s.shellPid)
-	if err != nil {
-		log.Errorf("session %s, shell pid %d, find process error: %s", s.id, s.shellPid, err.Error())
-		return err
-	}
-	err = p.Signal(syscall.SIGINT)
-	if err != nil {
-		log.Errorf("session %s, shell pid %d, signal error: %s", s.id, s.shellPid, err.Error())
-		return err
-	}
-	s.pseudoTTY.Close()
+		if msg.Header.Proto == ws.ProtoTypeControl {
+			if sess.HandleControl(msg) {
+				return
+			}
+			continue
+		}
 
-	err = procps.TerminateAndWait(s.shellPid, s.command, 2*time.Second)
```

Comment:
just for my peace of mind (sorry, again) where did that go?

Reply:
`session/shell.go`